test/com/ucm/dasi/catan/warehouse/WarehouseTest.java

```diff
@@ -0,0 +1,103 @@
+package com.ucm.dasi.catan.warehouse;
+
+import static org.junit.Assert.assertEquals;
+import java.util.Map;
+import org.junit.Test;
+import com.ucm.dasi.catan.warehouse.exception.NegativeNumberException;
+import com.ucm.dasi.catan.warehouse.exception.NotEnoughtResourcesException;
+
+public class WarehouseTest {
+
+	public WarehouseTest() {
+
+	}
+
+	/* CHECK QUANTITY TESTS */
+
+	@Test
+	public void checkQuantityNoEmptyConstructor() throws NegativeNumberException {
+
+		Map<ResourceType, Integer> resources = Map.of(ResourceType.Ore, 2, ResourceType.Brick, 2, ResourceType.Wool, 2,
+				ResourceType.Lumber, 2, ResourceType.Gain, 2);
+
+		Warehouse warehouse = new Warehouse(resources);
+		assertEquals(warehouse.getQuantityResource(), 10, 0);
+
+	}
+
+	@Test
+	public void checkQuantityEmptyConstructor() {
+		Warehouse warehouse = new Warehouse();
+		assertEquals(warehouse.getQuantityResource(), 0, 0);
+	}
+
+	@Test
+	public void checkQuantityAdd() throws NegativeNumberException {
+
+		Map<ResourceType, Integer> resources = Map.of(ResourceType.Ore, 2, ResourceType.Brick, 2, ResourceType.Wool, 2,
+				ResourceType.Lumber, 2, ResourceType.Gain, 2);
+
+		Warehouse warehouse = new Warehouse();
+		warehouse.add(new Warehouse(resources));
+		assertEquals(warehouse.getQuantityResource(), 10, 0);
+	}
+
+	@Test
+	public void checkQuantitySubstract() throws NegativeNumberException, NotEnoughtResourcesException {
+
+		Map<ResourceType, Integer> resources1 = Map.of(ResourceType.Ore, 2, ResourceType.Brick, 2, ResourceType.Wool, 2,
+				ResourceType.Lumber, 2, ResourceType.Gain, 2);
+
+		Map<ResourceType, Integer> resources2 = Map.of(ResourceType.Ore, 2, ResourceType.Brick, 2, ResourceType.Wool, 3,
+				ResourceType.Lumber, 2, ResourceType.Gain, 2);
+
+		Warehouse warehouse1 = new Warehouse(resources1);
+		Warehouse warehouse2 = new Warehouse(resources2);
+		warehouse2.substract(warehouse1);
+
+		assertEquals(warehouse2.getQuantityResource(), 1, 0);
+
+	}
+
+	@Test
+	public void checkQuantityAddSubstract() throws NegativeNumberException, NotEnoughtResourcesException {
```

Comment:
We code unit tests. This test is already covered by the others and must be removed.

Reply:
Thats not true, I tryed add and substract in different methods, but not both in the same. It's better to have more tests to prove that the logic is write rather than be stingy.

Reply:
Nope, please consider reading this brief [introduction](http://softwaretestingfundamentals.com/unit-testing/). Unit tests do __not__ test multiple operations, just one.

This tests reduces the maintainability rating of the project

Reply:
Yes, if we understand that this warehouse is a "individual units/ components of a software". This documentation give me the reason.

Reply:
More if this two methods can break the functionality of each other. Be stingy in this kind of things it's not a good idea.

Reply:
An operation is a public method. A unit test must __not__ cover two public methods. It can use multiple public methods but must __not__ cover them.

Reply:
> More if this two methods can break the functionality of each other. Be stingy in this kind of things it's not a good idea.

Please could you provide an example?

Reply:
There are hundreds. For example. Maybe the add method can break the quantityResource variable, doing the operation wrong and that makes that subtract method thinks that there are fewer resources, and for example thief appears with a 7 in the dices, and the player quanqityResources indicates that don't have more than 7 resources and in reality he has more than 7 resources. Easy, and there are a lot of examples more.

Reply:
For me a class implements a functionality. And we can divide this in sub-functionalities, and this one have relationships between them. That's why it's important to check that one sub-functionality does not break the logic of others.

Reply:
> There are hundreds. For example. Maybe the add method can break the quantityResource variable, doing the operation wrong and that makes that subtract method thinks that there are fewer resources, and for example thief appears with a 7 in the dices, and the player quanqityResources indicates that don't have more than 7 resources and in reality he has more than 7 resources. Easy, and there are a lot of examples more.

In the case you provided a __good__ unit test battery would discover the bug. The method that test the add method should discover that bug.

You see you know hundreds, I just need one, but consider the following thoughts:

  - In a single threaded environment like the current one, only one action happens at a time.
  - An operation updates the state of an object.

So, given a set of operations, it makes no sense to try to cover all the operations:

- Given the operations O = [O1, O2 ... On]
- Given the initial state S

Covering the operations O over S is equal to:

  if O'(S) is appliyng all the operations in O' at a certain order
 - For each operations O' as a subset of O, cover O'(S).
 - But if O' is equal to [O'1, O'2 ... O'n] and O'' is equal to [O'1, O'2 ...O'(n-1)], covering O'(S) is equal to cover O'n(O''(S)), so unit testing is still a good choice

Reply:
> For me a class implements a functionality. And we can divide this in sub-functionalities, and this one have relationships between them. That's why it's important to check that one sub-functionality does not break the logic of others.

Please read my previous comment

Reply:
The objective of the Uni-tests are not to be optimal. Uni-tests are for check the functionality of our software. If the software needs, we need to do it. And about the first part, of course now we will no need concurrence in our software, but, we don't know if in the future we will have this case.

Reply:
> The objective of the Uni-tests are not to be optimal. Uni-tests are for check the functionality of our software. If the software needs, we need to do it. And about the first part, of course now we will no need concurrence in our software, but, we don't know if in the future we will have this case.

Please read the comment carefully I think you did not understand it.

We are not talking about performance, we are talking about the complexity of the test battery. We write unit test because we will write less tests and the test code will be more maintainable. My last comment was to explain we don't need to cover multiple cases since unit testing must test an operation against any possible state of the object.

We can start talking about concurrence if you want. The answer is: We take the risk.

- First of all, we would need `N!` cover a class with `N` public methods covering combos of methods.
- In addition, if we have a concurrent programming model, then we can unit test everything. Then we can add more test methods to cover concurrency conditions, but I won't let this project LOCs grow in `N!`.

Reply:
First of all, as i said, there are hundreds of examples that breaks the sub-functionality of each other and check this things is easy and obviously i will not take this kind of risks only to be stingy, if you want it, okey, but do it in your develops, no in mines. And another time, as i said, Uni-tests is to check that all works fine, if the difference between cover all sub-functionalities and not cover it is 0.2 ms, i will do it, of course.

Reply:
> First of all, as i said, there are hundreds of examples that breaks the sub-functionality of each other and check this things is easy and obviously i will not take this kind of risks only to be stingy, if you want it, okey, but do it in your develops, no in mines. And another time, as i said, Uni-tests is to check that all works fine, if the difference between cover all sub-functionalities and not cover it is 0.2 ms, i will do it, of course.

Could you read [this](https://github.com/notaphplover/catan-heroes/pull/1/files#r381404109) comment? You are not taking any risk at all. The dialog makes sense if we listen to each other